Mathlib/Data/Bitvec/Defs.lean

```diff
@@ -1,355 +1,239 @@
-protected def ofInt : ∀ n : ℕ, Int → Bitvec n
-  | n, Int.ofNat m => Bitvec.ofNat n m
-  | n, Int.negSucc m => (Bitvec.ofNat n m).not
+open Nat
 
-/-- `add_lsb r b` is `r + r + 1` if `b` is `true` and `r + r` otherwise. -/
-def addLsb (r : ℕ) (b : Bool) :=
-  r + r + cond b 1 0
-#align bitvec.add_lsb Bitvec.addLsb
+lemma testBit_eq_ofNat {x: BitVec w} :
+  Bool.toNat (testBit (x.val) k) = (BitVec.ofNat 1 (x.val >>> k)).val:= by
+  simp only [BitVec.ofNat, Fin.ofNat', testBit, shiftRight_eq_shiftr, mod_two_of_bodd, pow_one]
+  aesop
 
-/-- Given a `List` of `Bool`s, return the `nat` they represent as a list of binary digits. -/
-def bitsToNat (v : List Bool) : Nat :=
-  v.foldl addLsb 0
-#align bitvec.bits_to_nat Bitvec.bitsToNat
+lemma val_to_ofNat (h: m < 2^w) : (BitVec.ofNat w m).val = m := by
+  simp [BitVec.ofNat, Fin.ofNat', mod_eq_of_lt h]
 
-/-- Return the natural number encoded by the input bitvector -/
-protected def toNat {n : Nat} (v : Bitvec n) : Nat :=
-  bitsToNat (toList v)
-#align bitvec.to_nat Bitvec.toNat
+lemma ofNat_to_val (x : BitVec w) : BitVec.ofNat w x.val = x := by
+  simp [BitVec.ofNat, Fin.ofNat', mod_eq_of_lt x.isLt]
 
-instance (n : ℕ) : Preorder (Bitvec n) :=
-  Preorder.lift Bitvec.toNat
+lemma ofNat_to_val' (x : BitVec w) (h : v = w):
+  HEq x (BitVec.ofNat v x.val) := h ▸ heq_of_eq (ofNat_to_val x).symm
 
-/-- convert `fin` to `Bitvec` -/
-def ofFin {n : ℕ} (i : Fin <| 2 ^ n) : Bitvec n :=
-  Bitvec.ofNat _ i.val
-#align bitvec.of_fin Bitvec.ofFin
+theorem concat_ext {x : BitVec a} {y : BitVec b} :
+  (x ++ y).val = x.val <<< b ||| y.val := rfl
 
-/-- convert `Bitvec` to `fin` -/
-def toFin {n : ℕ} (i : Bitvec n) : Fin (2 ^ n) :=
-  i.toNat
-#align bitvec.to_fin Bitvec.toFin
+theorem extract_ext : (extract i j x).val = x.val/2^j % (2^(i - j + 1)) := by
+  simp [extract, BitVec.ofNat, Fin.ofNat', shiftRight_eq_shiftr, shiftr_eq_div_pow]
 
+theorem testBit_eq_rep {x: BitVec w} (i : Nat) (h: i< w): x[i] = testBit x.val i := by rfl
 
-/-- Return the integer encoded by the input bitvector -/
-protected def toInt : ∀ {n : Nat}, Bitvec n → Int
-  | 0, _ => 0
-  | succ _, v =>
-    cond (head v) (Int.negSucc <| Bitvec.toNat <| Bitvec.not <| tail v)
-      (Int.ofNat <| Bitvec.toNat <| tail v)
-#align bitvec.to_int Bitvec.toInt
+theorem testBit_eq_rep' {x: Nat} (i : Nat) (h: i< w) (h2: x< 2^w):
+  (BitVec.ofNat w x)[i] = testBit x i := by
+  simp [BitVec.ofNat, GetElem.getElem, lsbGet, extract, Fin.ofNat', mod_eq_of_lt, h2]
 
-end Conversion
+def bbT (bs : List Bool) : BitVec bs.length :=
+  ⟨ofBits (λ i => bs[i]!) 0 bs.length, @ofBits_lt _ (bs.length)⟩
 
-/-! ### Miscellaneous instances -/
 
+/-! ### Equivalence between bitwise and BitVec operations -/
 
-private def repr {n : Nat} : Bitvec n → String
-  | ⟨bs, _⟩ => "0b" ++ (bs.map fun b : Bool => if b then '1' else '0').asString
+theorem BV_add {x y : BitVec w}: bitadd x.val y.val w = (x + y).val := by
```

Comment:
I'm not 100% sure we want to have `bitadd` and related definitions on `Nat`; it seems to overlap quite a bit with the `Fin n -> Bool` alternative representation that was discussed on Zulip.

I would suggest limiting this PR to just the basic definitions on `Bitvec`tors, and removing all the extra definitions from `Nat.Bitwise`.

Reply:
Ok, I made that change. Now everything included is just necessary for definitions on ```BitVec```.